From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Mon, 8 Mar 2021 14:05:11 +0100
Subject: smartcard-manager: Ignore slots with invalid tokens

Some modules may ignore the fullSlots flags on C_GetSlotList, so do a
double check in such case instead of crashing.

Related-to: https://github.com/OpenSC/OpenSC/pull/2248
---
 plugins/smartcard/gsd-smartcard-manager.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/plugins/smartcard/gsd-smartcard-manager.c b/plugins/smartcard/gsd-smartcard-manager.c
index 3ec1bd2..cf59443 100644
--- a/plugins/smartcard/gsd-smartcard-manager.c
+++ b/plugins/smartcard/gsd-smartcard-manager.c
@@ -448,7 +448,17 @@ sync_initial_tokens_from_driver (GsdSmartcardManager *self,
 
         for (l = full_slots; l; l = l->next) {
                 GckSlot *slot = l->data;
-                GckTokenInfo *token_info = gck_slot_get_token_info (slot);
+                GckTokenInfo *token_info;
+
+                if (!gsd_smartcard_utils_slot_has_flags (slot, CKF_TOKEN_PRESENT)) {
+                        CK_FUNCTION_LIST_PTR p11k_module = gck_module_get_functions (module);
+
+                        g_warning ("Module %s returned slot with no tokens",
+                                   p11_kit_module_get_name (p11k_module));
+                        continue;
+                }
+
+                token_info = gck_slot_get_token_info (slot);
 
                 g_debug ("Detected smartcard '%s' in slot %lu at start up",
                          token_info->label, gck_slot_get_handle (slot));
